transport/http/handler_test.go

```diff
@@ -0,0 +1,190 @@
+
+	headersWithBadTTL := headerCopyWithout(baseHeaders, TTLMSHeader)
+	headersWithBadTTL.Set(TTLMSHeader, "not a number")
+
+	tests := []struct {
+		req *http.Request
+		msg string
+	}{
+		{&http.Request{Method: "GET"}, "not found"},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headerCopyWithout(baseHeaders, CallerHeader),
+			},
+			"caller name is required",
+		},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headerCopyWithout(baseHeaders, ServiceHeader),
+			},
+			"service name is required",
+		},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headerCopyWithout(baseHeaders, ProcedureHeader),
+			},
+			"procedure name is required",
+		},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headerCopyWithout(baseHeaders, TTLMSHeader),
+			},
+			"ttlms is required",
+		},
+		{
+			&http.Request{
+				Method: "POST",
+				Header: headersWithBadTTL,
+			},
+			"ttlms must be an int",
+		},
+	}
+
+	for _, tt := range tests {
+		req := tt.req
+		if req.Body == nil {
+			req.Body = ioutil.NopCloser(bytes.NewReader([]byte{}))
+		}
+
+		h := handler{transporttest.NewMockHandler(mockCtrl)}
+		rw := httptest.NewRecorder()
+		h.ServeHTTP(rw, tt.req)
+
+		code := rw.Code
+		assert.True(t, code >= 400 && code < 500, "expected 400 level code")
+		assert.Contains(t, rw.Body.String(), tt.msg)
+	}
+}
+
+func TestHandlerInternalFailure(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	headers := make(http.Header)
+	headers.Set(CallerHeader, "somecaller")
+	headers.Set(EncodingHeader, "raw")
+	headers.Set(TTLMSHeader, "1000")
+	headers.Set(ProcedureHeader, "hello")
+	headers.Set(ServiceHeader, "fake")
+
+	request := http.Request{
+		Method: "POST",
+		Header: headers,
+		Body:   ioutil.NopCloser(bytes.NewReader([]byte{})),
+	}
+
+	rpcHandler := transporttest.NewMockHandler(mockCtrl)
+	rpcHandler.EXPECT().Handle(
+		gomock.Any(),
+		transporttest.NewRequestMatcher(
+			t, &transport.Request{
+				Caller:    "somecaller",
+				Service:   "fake",
+				Encoding:  raw.Encoding,
+				TTL:       time.Second,
+				Procedure: "hello",
+				Body:      bytes.NewReader([]byte{}),
+			},
+		), gomock.Any(),
+	).Return(fmt.Errorf("great sadness"))
+
+	httpHandler := handler{rpcHandler}
+	httpResponse := httptest.NewRecorder()
+	httpHandler.ServeHTTP(httpResponse, &request)
+
+	code := httpResponse.Code
+	assert.True(t, code >= 500 && code < 600, "expected 500 level response")
+	assert.Contains(t, httpResponse.Body.String(), "great sadness")
+}
+
+func headerCopyWithout(headers http.Header, names ...string) http.Header {
+	blacklist := make(map[string]struct{}, len(names))
+	for _, n := range names {
+		blacklist[strings.ToLower(n)] = struct{}{}
+	}
+
+	newHeaders := make(http.Header)
+	for k, vs := range headers {
+		if _, ok := blacklist[strings.ToLower(k)]; ok {
+			continue
+		}
+		for _, v := range vs {
+			newHeaders.Add(k, v)
+		}
+	}
+
+	return newHeaders
+}
+
```

Comment:
Test for setting blank header, and getting a HTTP header that has a blank value (or a non-blank value followed by blank value)